adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/VertxBasedMqttProtocolAdapter.java

```diff
@@ -62,22 +81,117 @@ private void bindMqttServer(final Future<Void> startFuture) {
 
     }
 
+    private void connectToHono(final Handler<AsyncResult<HonoClient>> connectHandler) {
+
+        // make sure that we are not trying to connect multiple times in parallel
+        this.hono = HonoClient.HonoClientBuilder.newClient()
+                .vertx(vertx)
+                .name(NAME)
+                .host(honoClientConfig.getHost())
+                .port(honoClientConfig.getPort())
+                .user(honoClientConfig.getUsername())
+                .password(honoClientConfig.getPassword())
+                .build();
+        ProtonClientOptions options = new ProtonClientOptions()
+                .setReconnectAttempts(-1)
+                .setReconnectInterval(200); // try to re-connect every 200 ms
+        this.hono.connect(options, connectAttempt -> {
+            if (connectHandler != null) {
+                connectHandler.handle(connectAttempt);
+            }
+        });
+    }
+
     @Override
     public void start(Future<Void> startFuture) throws Exception {
 
         this.bindMqttServer(startFuture);
+        this.connectToHono(null);
     }
 
     @Override
     public void stop(Future<Void> stopFuture) throws Exception {
 
-        // TODO
+        Future<Void> shutdownTracker = Future.future();
+        shutdownTracker.setHandler(done -> {
+            if (done.succeeded()) {
+                LOG.info("MQTT adapter has been shut down successfully");
+                stopFuture.complete();
+            } else {
+                LOG.info("error while shutting down MQTT adapter", done.cause());
+                stopFuture.fail(done.cause());
+            }
+        });
+
+        Future<Void> serverTracker = Future.future();
+        if (this.server != null) {
+            this.server.close(serverTracker.completer());
+        } else {
+            serverTracker.complete();
+        }
+        serverTracker.compose(d -> {
+            if (this.hono != null) {
+                this.hono.shutdown(shutdownTracker.completer());
+            } else {
+                shutdownTracker.complete();
+            }
+        }, shutdownTracker);
     }
 
     private void handleEndpointConnection(final MqttEndpoint endpoint) {
 
         LOG.info("Connection request from client {}", endpoint.clientIdentifier());
 
-        endpoint.writeConnack(MqttConnectReturnCode.CONNECTION_ACCEPTED, false);
+        if (!this.isConnected()) {
+            endpoint.writeConnack(MqttConnectReturnCode.CONNECTION_REFUSED_SERVER_UNAVAILABLE, false);
+
+        } else {
+
+            endpoint.publishHandler(message -> {
+
+                LOG.info("Just received message on [" + message.topicName() + "] payload [" + message.payload().toString(Charset.defaultCharset()) + "] with QoS [" + message.qosLevel() + "]");
```

Comment:
this looks more like a debug message to me, also please use the string formatting capabilities of slf4j instead of using string concatenation